<launch>

	<!-- glob params -->
	<arg name="use_markers_camera"  default="true" />
	<arg name="use_realsense"  default="true" />
	<arg name="markers_camera_name" default="markers_camera" />
    <arg name="fps"              default="30" />
	<arg name="actuator" default="36" />
	<arg name="step_div" default="100" />
	<arg name="f_ctrl" default="0.5" /> 
	<arg name="debug" default="false" /> 
	<arg name="epochs" default="5" /> 

	<!-- detector params -->
	<arg name="marker_length" default="0.01" />
	<arg name="use_reconfigure" default="false" />
	<arg name="filter" default="none" doc="Filter marker pose detections. One of ['none', 'mean', 'median', 'kalman_simple', 'kalman']"/>
	<arg name="filter_iters" default="10" />
	<arg name="vis" default="true" />
	<arg name="use_aruco" default="false" />
	<arg name="plot_id" default="-1" />
	<arg name="rh8d_port" default="/dev/ttyUSB0" />
	<arg name="rh8d_baud" default="1000000" />
	<arg name="qdec_port" default="/dev/ttyUSB2" />
	<arg name="qdec_baud" default="19200" />
	<arg name="qdec_filter_iters" default="200" />

	<!-- rs params -->
	<arg name="serial_no"           default="822512060411"/>
	<arg name="enable_depth"        default="false"/>
	<arg name="depth_width"         default="1280"/>
	<arg name="depth_height"        default="720"/>  
	<arg name="depth_fps"             default="$(arg fps)"/>
	<arg name="enable_color"        default="true"/>
	<arg name="color_width"         default="1920"/>
	<arg name="color_height"        default="1080"/>
	<arg name="color_fps"           default="$(arg fps)"/>
	<arg name="filters"                   default="colorizer"/>
	<arg name="clip_distance"             default="-2"/>
	<arg name="align_depth"               default="false"/>
	<arg name="output"               default="log"/>

	<!-- webcam args -->
    <arg name="always_capture"                            default="true"                                       doc="If true, always capture camera images, even if no one is subscribed to the output topics yet"/> 
    <arg name="video_stream_provider"            default="/dev/video0"                         doc="video device"/> 
    <arg name="width"                                              default="1280"                                        doc="set resolution"/> 
    <arg name="height"                                            default="960"                                         doc="set resolution"/> 
    <arg name="camera_info_url"                          default="file:///$(find nicol_rh8d)/cfg/camcalib_$(arg width)x$(arg height).yaml"  doc=" camera info loading"/>

	<!-- common manager -->
	<node  if="$(arg use_markers_camera)" pkg="nodelet" type="nodelet" name="manager" args="manager" output="screen" required="true" />
	<!-- markers cam -->
	<include if="$(eval use_markers_camera and use_realsense)" file="$(find nicol_rh8d)/launch/realsense.launch">
		<arg name="external_manager"    value="true"/>
		<arg name="manager"             value="/manager"/>
		<arg name="use_realsense" value="$(arg use_markers_camera)" />
		<arg name="camera"             value="$(arg markers_camera_name)" />
		<arg name="serial_no"           value="$(arg serial_no)"/>
		<arg name="enable_depth"        value="$(arg enable_depth)"/>
		<arg name="depth_width"         value="$(arg depth_width)"/>
		<arg name="depth_height"        value="$(arg depth_height)"/>  
		<arg name="depth_fps"           value="$(arg depth_fps)"/>
		<arg name="enable_color"        value="$(arg enable_color)"/>
		<arg name="color_width"         value="$(arg color_width)"/>
		<arg name="color_height"        value="$(arg color_height)"/>
		<arg name="color_fps"           value="$(arg color_fps)"/>
		<arg name="filters"                   value="$(arg filters)"/>
		<arg name="clip_distance"             value="$(arg clip_distance)"/>
		<arg name="align_depth"               value="$(arg align_depth)"/>
		<arg name="output"               value="$(arg output)"/>
	</include>
	<include if="$(eval use_markers_camera and not use_realsense)" ns="$(arg markers_camera_name)/color"  file="$(find nicol_rh8d)/launch/webcam.launch">
		<arg name="external_manager"    value="true"/>
		<arg name="nodelet_manager"     value="/manager"/>
		<arg name="fps"                                   value="$(arg fps)" />
    	<arg name="always_capture"           value="$(arg always_capture)" />
    	<arg name="camera_name"              value="$(arg markers_camera_name)" />
    	<arg name="video_stream_provider"   value="$(arg video_stream_provider)" />
    	<arg name="width"                                value="$(arg width)" />
    	<arg name="height"                              value="$(arg height)" />
    	<arg name="camera_info_url"           value="$(arg camera_info_url)" />
	</include>

	<!-- dataset collector script -->
	<node pkg="nicol_rh8d" name="dataset_collector" type="detect.py" output="screen" required="true">
		<param name="markers_camera_name" value="$(arg markers_camera_name)/color" />
		<param name="marker_length" value="$(arg marker_length)" />
		<param name="use_reconfigure" value="$(arg use_reconfigure)" />
		<param name="vis" value="$(arg vis)" />
		<param name="filter" value="$(arg filter)" />
		<param name="filter_iters" value="$(arg filter_iters)" /> 
		<param name="f_ctrl" value="$(arg f_ctrl)" />
		<param name="use_aruco" value="$(arg use_aruco)" />
		<param name="plot_id" value="$(arg plot_id)" />
		<param name="rh8d_port" value="$(arg rh8d_port)" />
		<param name="rh8d_baud" value="$(arg rh8d_baud)" />
		<param name="qdec_port" value="$(arg qdec_port)" />
		<param name="qdec_baud" value="$(arg qdec_baud)" />
		<param name="qdec_filter_iters" value="$(arg qdec_filter_iters)" />
		<param name="actuator" value="$(arg actuator)" />
		<param name="step_div" value="$(arg step_div)" /> 
		<param name="epochs" value="$(arg epochs)" /> 
		<param name="debug" value="$(arg debug)" /> 
		<param name="qdec_detect" value="true" />
	</node>

</launch>
